### ISSUE TITLE
Suddenly stopped working?

### ISSUE THREAD

**Issue by 097115** [deanishe/alfred-glosbe#4]
Says `Error contacting Glosbe.com` even though [glosbe.com](https://glosbe.com/) is alive and kicking :)

**Comment by deanishe** [comment]
Not much I can do with that. What does Alfred's debugger and/or the log file say? What version of macOS are you running?

**Comment by 097115** [comment]
Run it outside of Alfred:

    ~/Desktop> ./translate.py eng rus table
    21:38:34 translate.py:133 DEBUG    script args : [u'eng', u'rus', u'table']
    21:38:34 translate.py:135 DEBUG    docopt args : {u'--help': False,
     u'--langs': False,
     u'--openhelp': False,
     u'--text': False,
     u'<dest>': u'rus',
     u'<query>': u'table',
     u'<source>': u'eng'}
    21:38:34 translate.py:60 DEBUG    Calling http://glosbe.com/gapi_v0_1/translate with args : {u'dest': u'rus', u'phrase': u'table', u'from': u'eng', u'pretty': u'true', u'format': u'json'}
    <?xml version="1.0" encoding="utf-8"?>
    <items><item valid="no"><title>Error contacting Glosbe.com</title><subtitle /><icon>/System/Library/CoreServices/CoreTypes.bundle/Contents/Resources/AlertStopIcon.icns</icon></item></items>21:38:35 workflow.py:970 DEBUG    Cached data saved at : /Users/me/Library/Caches/com.runningwithcrayons.Alfred-2/Workflow Data/net.deanishe.alfred-glosbe/d7d0d4c07379d65e7bcf6935ac264722.cache
    21:38:35 workflow.py:1265 ERROR    object of type 'int' has no len()
    Traceback (most recent call last):
      File "/Users/me/Desktop/workflow/workflow.py", line 1263, in run
        func(self)
      File "./translate.py", line 180, in main
        if not len(results):
    TypeError: object of type 'int' has no len()

**Comment by RyuX51** [comment]
I second that. Running macOS 10.14.3, Alfred 3.8, same error (`object of type 'int' has no len()`). Really would appreciate your time. I use your workflow practically all the time. :)

**Comment by 097115** [comment]
It really looks like the problem is on glosbe.com's side (their API return Error 429, too many requests, like [this](https://glosbe.com/gapi/translate?from=pol&dest=eng&format=json&phrase=witaj&pretty=true)), so I doubt there's much @deanishe can do at the moment :(

**Comment by deanishe** [comment]
Yeah, I've just tested the API with cURL. The problem is as Glosbe's end, so there's nothing I can do.

Sorry :(

**Comment by RyuX51** [comment]
I see. So let's hope that this is just temporary. Thanks, @097115. And thank you @deanishe for this awesome workflow.

**Comment by 097115** [comment]
Well, we can send a few emails to Glosbe, their address is listed in the Contacts section :)

**Comment by akaralar** [comment]
@097115 did you send an email and got back a response? I was using this workflow literally all the time, such a bummer that it's not working. 

@deanishe I love the fact that the workflow presents more than one translation for a given phrase, I hope that can be achieved with other services as well. How hard would it be to integrate this with another free translation service?

**Comment by deanishe** [comment]
> How hard would it be to integrate this with another free translation service?

That depends entirely on the service.

**Comment by 097115** [comment]
> did you send an email and got back a response?

And you? I did, yes, but with no response. They also got a [facebook page](https://www.facebook.com/Glosbe), you may try contacting them there.

> I was using this workflow literally all the time, such a bummer that it's not working.

It sure not a replacement but you can try adding simple Glosbe (or Google Translate for that matter) web search to Alfred (`+` button -> Templates -> Web and URLs).

**Comment by akaralar** [comment]
@097115 yeah I did but no response yet as well. Too bad the API is not working.

@deanishe I'm wondering whether it would be viable to `curl` a translation page like `https://glosbe.com/{fromLang}/{toLang}/{query}` and parse the resulting HTML into a format that was returned from the API. Do you by any chance have any example JSON that was returned from the API?

**Comment by deanishe** [comment]
> I'm wondering whether it would be viable to curl a translation page like `https://glosbe.com/{fromLang}/{toLang}/{query}` and parse the resulting HTML

It might be possible to do, but I'm generally leery of HTML parsing as it's so prone to breaking.

The is an autosuggest endpoint that returns JSON:

```
https://glosbe.com/ajax/phrasesAutosuggest?from=en&dest=de&phrase=testing
```

But that's just search suggestions, not search results…

**Comment by vitorgalvao** [comment]
I’m pretty sure it still wasn’t working last weekend but it works now.

**Comment by akaralar** [comment]
@vitorgalvao interesting, it still doesn't work for me, and when I try to get anything from the API I still get 429 too many requests. Did you edit the workflow in any way or does it work the way it is now for you?

**Comment by vitorgalvao** [comment]
@akaralar No modification. But it no longer works today.

**Comment by akaralar** [comment]
@deanishe I'll be moving to a new country in 2 months and am planning on using this workflow extensively (because ease of use :)), so I'm willing to take a shot at html parsing in the following weeks.  My python knowledge is very basic so I'll have a chance to practice as well. Any recommendations on how to go forward with it?

**Comment by vitorgalvao** [comment]
> My python knowledge is very basic (…) Any recommendations on how to go forward with it?

Start by considering not doing it in Python, but instead in whatever language you’re familiar with. The API is still in place, so changing this Workflow to HTML parsing is premature; a different one makes more sense for now.

If you still want to do it in Python (and even if you don’t) start by making a few queries on the website and understanding its structure. Is it even parseable consistently? Because some websites are not (hint: there’s a popular thesaurus website whose structure makes such a thing too hard).

Next, do a few `curl` requests to the page and `grep` the info you want. Is it even there? Maybe it’s populated via JavaScript, in which case this is a losing endeavour.

After that (and that should be first so as to not waste time with premature programming) check your language’s popular scraping packages — there should be a consensus on what to use.

In Ruby I’m partial to `nokogiri`. I’ve read good things about `beautifulsoup` in Python.

**Comment by akaralar** [comment]
@vitorgalvao thanks for the suggestions! It might very well be premature but it has been broken for more than a month and they don't reply to emails sent by me and @097115 , so it will be nice to have a fallback mechanism when the API does not work. It can continue to function with the API when it is available.

I was planning to do it python since the workflow is in python. I don't want to replace all the functionality, just the part that makes an API request, as duplicating all the functionality might take more than I'd like to spare for this. I'm most comfortable in Swift and Objective-C and I don't think the scraping libraries in those languages are as mature as the stuff in ruby/python/js ie languages of the web.

The points you raised are very valid. I'll do my due diligence and try to understand whether it's feasible before starting with it. I hope their website is parsable!

**Comment by vitorgalvao** [comment]
> It might very well be premature but it has been broken for more than a month

I don’t mean it’s premature to make another Workflow; I mean it’s premature to replace the functionality in *this* Workflow. Making another one, even if it uses elements from this one, makes more sense for now.

> I was planning to do it python since the workflow is in python. I don't want to replace all the functionality, just the part that makes an API request, as duplicating all the functionality might take more than I'd like to spare for this.

Makes sense.

> I hope their website is scraper-friendly!

Even if it’s not, by checking the web requests you may find a private API. That’s might be even better.

**Comment by nandekonst** [comment]
> Well, we can send a few emails to Glosbe, their address is listed in the Contacts section :)

I experienced the same thing "too many request". I have written several e-mails, but never got any response.

**Comment by deanishe** [comment]
Yeah, I think this workflow is dead…

I don’t really want to convert it to HTML scraping (or private APIs). That tends to break often, and as I don’t really use this workflow much myself any more, I’m not really willing to commit to supporting such a potentially brittle workflows.
